---
title: "Initial Data Analysis"
author: "Coleman"
date: "2023-07-19"
output: html_document
---

BiocManager::install("SNPlocs.Hsapiens.dbSNP150.GRCh38")

Using dbSNP data from [https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/](NCBI's website).


```{r setup, include=FALSE}
library(targets)
knitr::opts_chunk$set(echo = TRUE)

read_vcf <- function(ifile){
  # Wrapper around read function, returns DataFrame
  vcf <- readVcf(ifile, "hg38")
  vcf
}

# dbsnp.vcf <- read_vcf("../data/dbSNP/common_all_20180418.vcf.gz")
vcf.df <- read_vcf("../../data/variant-calls/253.snps.vcf")
methy.df <- fread("../../data/methylation-counts/253.bed")


```

# Are the SNP Calls Reasonable?

Quantity and quality are both of concern. How many novel SNPs does this person have? How many heterozygous?

To generate SNP calls, we used `bs_call` with a modification from previous iterations (when our needs were different).

```
# Standard bs_calling, but keep all positions (not just CpGs and CpHs)
samtools view -h -f 2 253.bam | bs_call --all-positions -r reference.fna  > 253.called.split.vcf

# Keep variants that pass all filters (set by GATK)
bcftools view -c1 -i "FILTER='PASS'" 253.all.called.vcf > 253.snps.vcf
```

```{r}
tbl <- table(geno(vcf.df)$GT)
# tbl_dat <- as.data.frame(tbl)
tbl
```

We have 2.4 million heterozygous loci, 1.3 million that are homozygous alt, and 1900 that are heterozygous alt.

# How close are the SNPs to CpG loci? 


```{r construct_granges_objects}
snp.loci <- rowRanges(vcf.df)
cpg.loci <- makeGRangesFromDataFrame(methy.df, 
                                     starts.in.df.are.0based = T, 
                                     ignore.strand = T)
```


```{r how_many_snps_near_cpg}

gap <- c(250, 500, 1000, 2000, 5000)

N.snps.near.CpGs <- rep(NA, length(gap))

for (ii in 1:length(gap)){
  N.snps.near.CpGs[ii] <- length(unique(subsetByOverlaps(snp.loci, cpg.loci, maxgap = gap[ii])))
}

tmp.df <- data.frame(N.unique.SNPs = N.snps.near.CpGs, Width = gap*2)

```

```{r}
tmp.df %>% 
  ggplot(aes(x = Width, y = N.unique.SNPs)) +
  geom_point() +
  theme_minimal() +
  ggtitle("Nearly all SNPs are within a few kb of a CpG") +
  geom_hline(yintercept = length(vcf.df), color = "red") +
  xlab("Width of interval around CpG") +
  ylab("Number of distinct SNPs within interval") 
```
# How Many SNPs are Novel versus Common?

Hard to answer on the machine. Can follow up on remote machine at a later date.

# Best way to process this data?

Snake make pipeline

- Use bedtools, get fasta. Methylated counts file gives CpG loci (need to remove header).
  - Script to "expand BED" by some amount (e.g. 1000)
  - Run on a small window first to make sure we're actually centered on CpG
- Then, after binning with [bedtools](https://bedtools.readthedocs.io/en/latest/content/tools/getfasta.html), we perform the one-hot encoding. Then swap out as needed...
  - Filter down to intervals with at least one SNP
  - Then 